src/librustc_typeck/check/closure.rs

```diff
@@ -269,4 +273,408 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+/// - Second, any bound region appears that appears in some expected type E,
+///   also appears in the corresponding spot in the resulting type R.
+///   - In particular, imagine that we have an expected signature
+///     `(&str) -> &str`, but the user wrote `|x: &'a str| ...`. In
+///     the supplied type, the region `'a` is free.
+///     - We could choose to use the final signature `(&'a str) ->
+///       &str` without violating the first invariant, since we did
+///       "intermix" any bound region names from the supplied types
+///       with those of the expected type.
+///     - However, if we did so, we have lost the constraint expressed
+///       in the expected signature, that the return type must relate
+///       to the input.  In fact, this resulting signature is
+///       inconsistent, since a bound region cannot only appear in the
+///       input (this was #38714).
+///
+/// To produce the type R, we relate the types recursively, which
+/// requires that the types generally match. If that is not true, an
+/// error results, which will result in the expected signature being
+/// ignored. The places where the types do not have to match, and the
+/// resulting type in those scenarios, are as follows:
+///
+/// - If the expected type E is an (unbound) inference variable,
+///   and S is not an inference variable, then the resulting type will be S,
+///   so long as S does not contain anonymous regions.
+/// - If the supplied type S is an unbound inference variable (i.e.,
+///   the user wrote `_`), we will use E.
+/// - When we encounter types with internal binders, e.g. `fn(&str)`
+///   or `dyn Fn(&str)`, then we want to take the supplied type
+///   (justification below). To preserve the invariants listed above,
+///   we will error if either of the following is true (but otherwise
+///   use the supplied type):
+///   - the supplied type includes any regions bound at the closure depth
+///     - in this case, using the supplied type would inject names from the supplied
+///       side into the closure binder
+///   - the expected type includes any regions bound at the closure depth
+///     - in this case, using the supplied type would **remove** names from the closure
+///       binder that the expected type included
+///
+/// Regions are allowed to differ as follows:
+///
+/// - If S supplies a late-bound region bound at the closure depth (1):
+///   - If that region is anonymous, we use the region from E (which may also be free).
+///   - Otherwise, we yield an error. Note that this case is not possible with Rust's
+///     present syntax, it would require something like `foo(for<'a> |x: &'a u32| ...)`.
+/// - If both E and S supply a free region, we use the region from S (which,
+///   due to the limits of Rust's syntax -- must incidentally be a named region).
+/// - If E supplies a bound region but S does not, we error out.
+///
+/// # Why we use the supplied type with bound regions
```

Comment:
We'll force the expected type to be equal to the supplied type anyway when we resolve the "closure: expectation" obligation. so we always want to use the supplied type except for bound regions.

Reply:
I don't understand this comment. Keep in mind that the output of this algorithm is the ultimate type T which is assigned to the closure -- it may not be equal to what I've been calling the "supplied type", which is the type that the user actually wrote, at least as presently interpreted (in particular with respect to anonymous regions, which may wind up being mapped to other things).

It is this ultimate type T that will be equated with the expected type.

In other words, we never force the expected type to be equal to the supplied type.